proposals/0303-swiftpm-extensible-build-tools.md

```diff
@@ -432,77 +470,106 @@ protocol Path: ExpressibleByStringLiteral, CustomStringConvertible {
+Examples of plugins that need to use prebuild commands include SwiftGen and other tools that need to see all the input files, and whose set of output files is determined by the *contents* (not just the paths) of the input files. Such a plugin usually generates just one command and configures it with an output directory into which all generated sources will be written.
 
-Prebuild commands should be used when the tool being invoked can produce outputs whose names are determined by the _contents_ (not names) of the input files, or when there are other reasons why the names of the outputs cannot be known before actually running the command.
+Because they run on every build, prebuild commands should use caching to do as little work as possible (ideally none) when there are no changes to their inputs. This should include preserving timestamps of generated source files that haven't changed, since most build systems (including SwiftPM's own build system) use timestamps as a shorthand for detecting whether files have changed.
+
+##### Build Commands
+
+Commands created using `addBuildCommand()` are incorporated into the build system's dependency graph, so that they run as needed during the build, based on their declared inputs and outputs. This requires that the paths of any outputs can be known before the command is run. This is usually done by forming the names of the outputs from some combination of the output directory and the name of the input file.
+
+Examples of plugins that can use regular build commands include compiler-like translators such as Protobuf and other tools that take a fixed set of inputs and produce a fixed set of outputs. (note that one nuance with Protobuf in particular is that it is actually up to the source generator invoked by `protoc` to determine the output paths — however, the relevant source generators for Swift and C do produce output files with predictable names).
+
+Regular build commands with defined outputs are preferable whenever possible, because such commands don't have to run unless their outputs are missing or their inputs have changed since the last time they ran.
+
+##### Postbuild Commands
+
+Commands created using `addPostbuildCommand()` are run after every build. They have no input or output directories, since they provide no inputs to the build.
+
+Instead, they can be passed information about whether the build succeeded or failed, along with other information about the build. SwiftPM vends this information via a JSON file whose path is provided in the `postbuildResultFile` property of the target build context passed to the plugin. The plugin may choose to pass this path to the postbuild command in any way that it wants to, such as on the command line or in an environment variable.
```

Comment:
Oh I see... so you're trying to make it easy for non swift tools to get hold of it too; or rather, for external process based ones?

Reply:
This is because of the distinction between a plugin and the command it invokes.  The plugins are evaluated after the package graph is, and they can create prebuild commands, regular build command, and postbuild commands.  The postbuild commands are run after the build and so they need to be passed the path of the results file.  The plugin is the switchboard between SwiftPM and the command, by passing the path to the command.  It doesn't matter what language the command is implemented in.